how to register a rely function Issue-478 (#579)
microsoft/autogen: website/docs/FAQ.md

2. `mkdir /home/vscode/.local/lib/python3.10/site-packages/google/colab`

Explanation: Per [this gist](https://gist.github.com/defulmere/8b9695e415a44271061cc8e272f3c300?permalink_comment_id=4711478#gistcomment-4711478), linked from the official [chromadb docs](https://docs.trychroma.com/troubleshooting#sqlite), adding this folder triggers chromadb to use pysqlite3 instead of the default.

## How to register a reply function

(from [issue #478](https://github.com/microsoft/autogen/issues/478))

> You can hook in a function that is called whenever an agent receives a message

Reviewer: The registered reply function is only called when `generate_reply` is called. If an agent receive a message but a reply is not requested, it won't be called.

> Assuming you have two agents user_proxy and assistant, you can register a function with these agents that gets called with some parameters.

Reviewer 2: "you can register a function with these agents that gets called with some parameters" is a bit too generic on its own. Perhaps link to the documentation: https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent#register_reply

Author: Thanks @afourney

Reviewer: I agree with adding a link to the documentation.

```
def print_messages(recipient, messages, sender, config):
    if "callback" in config and  config["callback"] is not None:
        callback = config["callback"]
        callback(sender, recipient, messages[-1])
    print(f"Messages sent to: {recipient.name} | num messages: {len(messages)}")
    return False, None  # required to ensure the agent communication flow continues

user_proxy.register_reply(
    [autogen.Agent, None],
    reply_func=print_messages,
    config={"callback": None},
)

assistant.register_reply(
    [autogen.Agent, None],
    reply_func=print_messages,
    config={"callback": None},
)

```

> In the above, we register a print_messages message that is called each time the agent receive a message.

Reviewer: This is not accurate. See my comment above.

Explanation: Per [this gist](https://gist.github.com/defulmere/8b9695e415a44271061cc8e272f3c300?permalink_comment_id=4711478#gistcomment-4711478), linked from the official [chromadb docs](https://docs.trychroma.com/troubleshooting#sqlite), adding this folder triggers chromadb to use pysqlite3 instead of the default.

## How to register a reply function

> (from [issue #478](https://github.com/microsoft/autogen/issues/478))

Reviewer: What about copying the example from #478 to here?
@victordibia @gagb what do you think?

Reviewer 3: I think directly show the example would be good for readers

Author: Example added

Explanation: Per [this gist](https://gist.github.com/defulmere/8b9695e415a44271061cc8e272f3c300?permalink_comment_id=4711478#gistcomment-4711478), linked from the official [chromadb docs](https://docs.trychroma.com/troubleshooting#sqlite), adding this folder triggers chromadb to use pysqlite3 instead of the default.

## How to register a reply function

(from [issue #478](https://github.com/microsoft/autogen/issues/478))

See here https://microsoft.github.io/autogen/docs/reference/agentchat/conversable_agent/#register_reply

> For example,
> 
>  Assuming you have two agents user_proxy and assistant, you can register a function with these agents that gets called with some parameters. (The registered reply function is only called when `generate_reply` is called)

Reviewer: ```suggestion
For example, you can register a reply function that gets called when `generate_reply` is called for an agent.
```

```python
def print_messages(recipient, messages, sender, config):
    if "callback" in config and  config["callback"] is not None:
        callback = config["callback"]
        callback(sender, recipient, messages[-1])
    print(f"Messages sent to: {recipient.name} | num messages: {len(messages)}")
    return False, None  # required to ensure the agent communication flow continues

user_proxy.register_reply(
    [autogen.Agent, None],
    reply_func=print_messages,
    config={"callback": None},
)

assistant.register_reply(
    [autogen.Agent, None],
    reply_func=print_messages,
    config={"callback": None},
)
```

> In the above, we register a print_messages message that is called each time the agent receive a message.

Reviewer 3: print_messages message -> `print_messages` function

Author: fixed @JieyuZ2

> In the above, we register a `print_messages` function that is called each time the agent receive a message.

Reviewer: ```suggestion
In the above, we register a `print_messages` function that is called each time the agent's `generate_reply` is triggered after receiving a message.
```

Author: changes made
